resources/asciidoctor/spec/copy_images_spec.rb

```diff
@@ -0,0 +1,225 @@
+require 'copy_images/extension'
+require 'fileutils'
+require 'tmpdir'
+
+RSpec.describe CopyImages do
+  before(:each) do
+    Extensions.register do
+      tree_processor CopyImages
+    end
+  end
+
+  after(:each) do
+    Extensions.unregister_all
+  end
+
+  private
+  def copy_attributes copied
+    {
+      'copy_image' => Proc.new { |uri, source|
+        copied << [uri, source]
+      }
+    }
+  end
+
+  spec_dir = File.dirname(__FILE__)
+
+  it "copies a file when directly referenced" do
+    copied = []
+    attributes = copy_attributes copied
+    input = <<~ASCIIDOC
+      == Example
+      image::resources/copy_images/example1.png[]
+    ASCIIDOC
+    expect { convert input, attributes }.to raise_error { |error|
+      expect(error).to be_a(ConvertError)
```

Comment:
I found this testing idiom unintuitive. By description this appears to be a positive test, one which should do a thing (and behave doing so) and then verify the thing has been done. But this is looking for an error, and then that error is actually an INFO... it's got a lot to tease apart just to line up what it says it does with how it checks doing it.

I don't know yet if I am requesting a change or not, but even if I figure there's probably a really good explanation for it, I'd still like to hear it because it definitely made me go "...huh?" reviewing it.

Reply:
It exists because the tests throw a `ConvertError` if we log *anything* during the conversion. That way when we're testing that we don't log anything we don't have to do anything - we get that behavior by not catching the exception. It made more sense when I was catching warnings to be honest.

I think I can make this more clear. I'll give it a shot on Monday.